msal/tests/Test.MSAL.NET.Unit/RequestsTests/DeviceCodeRequestTests.cs

```diff
@@ -0,0 +1,198 @@
+    {
+        private TokenCache _cache;
+        private readonly MyReceiver _myReceiver = new MyReceiver();
+
+        [TestInitialize]
+        public void TestInitialize()
+        {
+            RequestTestsCommon.InitializeRequestTests();
+            Telemetry.GetInstance().RegisterReceiver(_myReceiver.OnEvents);
+
+            _cache = new TokenCache();
+        }
+
+        [TestCleanup]
+        public void TestCleanup()
+        {
+            _cache.tokenCacheAccessor.AccessTokenCacheDictionary.Clear();
+            _cache.tokenCacheAccessor.RefreshTokenCacheDictionary.Clear();
+        }
+
+        private HttpResponseMessage CreateDeviceCodeResponseSuccessMessage()
+        {
+            const string SuccessMessage = 
+                "{\"user_code\":\"B6SUYU5PL\"," +
+                 "\"device_code\":\"BAQABAAEAAADXzZ3ifr-GRbDT45zNSEFEfU4P-bZYS1vkvv8xiXdb1_zX2xAcdcfEoei1o-t9-zTB9sWyTcddFEWahP1FJJJ_YVA1zvPM2sV56d_8O5G23ti5uu0nRbIsniczabYYEr-2ZsbgRO62oZjKlB1zF3EkuORg2QhMOjtsk-KP0aw8_iAA\"," +
+                 "\"verification_url\":\"https://microsoft.com/devicelogin\"," +
+                 "\"expires_in\":\"900\"," +
+                 "\"interval\":\"5\"," +
+                 "\"message\":\"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code B6SUYU5PL to authenticate.\"}";
+
+            return MockHelpers.CreateSuccessResponseMessage(SuccessMessage);
+        }
+
+        [TestMethod]
+        [TestCategory("DeviceCodeRequestTests")]
+        public void TestDeviceCodeAuthSuccess()
+        {
+            var parameters = CreateAuthenticationParametersAndSetupMocks(out HashSet<string> expectedScopes);
+
+            // Check that cache is empty
+            Assert.AreEqual(0, _cache.tokenCacheAccessor.AccessTokenCacheDictionary.Count);
+            Assert.AreEqual(0, _cache.tokenCacheAccessor.AccountCacheDictionary.Count);
+            Assert.AreEqual(0, _cache.tokenCacheAccessor.IdTokenCacheDictionary.Count);
+            Assert.AreEqual(0, _cache.tokenCacheAccessor.RefreshTokenCacheDictionary.Count);
+
+            DeviceCodeResult actualDeviceCodeResult = null;
+            DeviceCodeRequest request = new DeviceCodeRequest(parameters, result => 
+            {
+                actualDeviceCodeResult = result;
+                return Task.FromResult(0);
+            });
+            var task = request.RunAsync(CancellationToken.None);
+            task.Wait();
+            var authenticationResult = task.Result;
+            Assert.IsNotNull(authenticationResult);
+            Assert.IsNotNull(actualDeviceCodeResult);
+
+            Assert.AreEqual("client_id", actualDeviceCodeResult.ClientId);
+            Assert.AreEqual("BAQABAAEAAADXzZ3ifr-GRbDT45zNSEFEfU4P-bZYS1vkvv8xiXdb1_zX2xAcdcfEoei1o-t9-zTB9sWyTcddFEWahP1FJJJ_YVA1zvPM2sV56d_8O5G23ti5uu0nRbIsniczabYYEr-2ZsbgRO62oZjKlB1zF3EkuORg2QhMOjtsk-KP0aw8_iAA", actualDeviceCodeResult.DeviceCode);

```

Comment:
>"BAQABAAEA [](start = 28, length = 10)

did you consider making this one and the usercode consts? perhaps even the url?

Reply:
#fixed